### PR TITLE
Icon

### PR DESCRIPTION
**Pull request by tienvx**
Provide a simple function for developer to render icon.
- `at_icon($name, $source)`
- Support fontawesome
- Support fontello
- Provide `drush fontello` to build fontello
- Two unit-test cases, one for fontawesome, one for fontello
### Ref
- https://github.com/v3knet/at_base/blob/route-editable/at_base.module#L227
- https://github.com/v3knet/at_base/blob/route-editable/config/require.yml#L30

### CODE REVIEW

File: at_base.drush.inc [hunk 1]
```diff
@@ -4,6 +4,7 @@ require_once dirname(__FILE__) . '/at_base.module';
 require_once dirname(__FILE__) . '/drush/at_services.inc';
 require_once dirname(__FILE__) . '/drush/at_cache.inc';
 require_once dirname(__FILE__) . '/drush/at_twig.inc';
+require_once dirname(__FILE__) . '/drush/fontello.inc';
```
Comment (and1truong) [inline at_base.drush.inc]: We should use same style, at_fontello.inc

File: config/require.yml [hunk 2]
```diff
@@ -4,7 +4,9 @@ projects:
     download:
       type: git
       url: 'https://github.com/mustangostang/spyc.git'
-      branch: '0.5.1'
+      branch: 'master'
```
Comment (and1truong) [inline config/require.yml]: Master maybe changed, our patch no longer applicable. I think we should use `eba310c26819a4fe519a60309e5d3994588c7840`

File: lib/Drupal/at_base/Tests/Unit/BreadcrumbTest.php [hunk 3]
```diff
@@ -8,7 +8,7 @@
 class BreadcrumbTest extends UnitTestCase {
   private $api;
 
-  public function getInfo() {
+  static public function getInfo() {
```
Comment (and1truong) [inline lib/Drupal/at_base/Tests/Unit/BreadcrumbTest.php:11]: Why?

**Comment by and1truong** [comment]
Can we have the simple function `at_icon($name, $source);`?

``` php
namespace Drupal\at_base {
  class Icon {
    protected $css;
    protected $tag = 'i';
    protected $class;
    protected $text;

    public function render() {
      !empty($this->css) && at_fn::drupal_add_css();
      return "<{$this->tag} class=\"{$this->class}\">{$this->text}</$this->tag>";
    }
  }
};

/**
 * Shortcut to render to icon.
 */
function at_icon($name, $source = 'icon.fontawesome') {
  try {
    /** @type $service Drupal\at_base\SourceInterface **/
    $service = at_container($source);

    /** @type $icon Drupal\at_base\Icon **/
    $icon = $service->get($name);
    return $icon->render();
  }
  catch (Exception $e) {
    return $e->getMessage();
  }
}
```

**Comment by and1truong** [comment]
Your spyc patch is already merged, the patch no longer need to by defined here.

File: lib/Drupal/at_base/Tests/Unit/BreadcrumbTest.php (hunk 3 above)
Comment (tienvx) [inline lib/Drupal/at_base/Tests/Unit/BreadcrumbTest.php:11]: Simpletest will show a warning, or even an error.

This is the warning:
<code>call_user_func() expects parameter 1 to be a valid callback, non-static method getInfo() should not be called statically'.</code>

File: lib/Drupal/at_base/Tests/Unit/FontAwesomeTest.php [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    $this->service = at_container('icon.fontawesome');
+
+    drupal_static_reset('fontawesome_library_added');
+  }
+
+  public function testHtmlGeneration() {
+    $css_code = 'fa-camera-retro';
+    $expected_html = '<i class="fa '. $css_code .'"></i>';
+
+    $icon = $this->service->get($css_code);
```
Comment (and1truong) [inline lib/Drupal/at_base/Tests/Unit/FontAwesomeTest.php]: Can we use at_icon() directly here?

File: lib/Drupal/at_base/Tests/Unit/FontElloTest.php [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    // Unicode character is cached.
+    $this->assertTrue(strpos($char, '\\') === 0, 'Unicode character is cached.');
+  }
+
+  public function testCssAdded() {
+
+    $css_code = 'github';
+
+    // Fake some values.
+    $font_path = 'sites/all/libraries/fontello/src/cool.font/font';
```
Comment (and1truong) [inline lib/Drupal/at_base/Tests/Unit/FontElloTest.php]: This should be `at_library('fontello') . 'src/cool.font/font'`

File: lib/Drush/Command/FontEllo.php [hunk 6]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+namespace Drupal\at_base\Drush\Command;
+
+class FontEllo {
+
+  public function execute() {
+    $fontello_library_path = at_library('fontello', NULL, FALSE);
+    $drupal_root = drush_get_context('DRUSH_DRUPAL_ROOT');
+//    $site_root = drush_get_context('DRUSH_DRUPAL_SITE_ROOT');
```
Comment (and1truong) [inline lib/Drush/Command/FontEllo.php]: Just remove it (:

File: lib/Drush/Command/FontEllo.php [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+    drush_shell_exec_interactive('cd ' . $drupal_root . '/' . $fontello_library_path. '&& cd .. && rm -rf fontello && git clone https://github.com/fontello/fontello.git');
```
Comment (and1truong) [inline lib/Drush/Command/FontEllo.php]: I think we should check fontello diretory must be a working git repository, throw exception if it's not.

File: lib/EmptyIcon.php [hunk 8]
```diff
@@ -0,0 +1,12 @@
+<?php
+
+namespace Drupal\at_base;
```
Comment (and1truong) [inline lib/EmptyIcon.php]: This should be `Drupal\at_base\Icon`

File: lib/Icon.php [hunk 9]
```diff
@@ -0,0 +1,29 @@
+<?php
+
+namespace Drupal\at_base;
```
Comment (and1truong) [inline lib/Icon.php]: This should be `Drupal\at_base\Icon`

File: lib/Icon/FontEllo.php [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+
+  /**
+   * Translate icon name to unicode character.
+   *
+   * @param type $name
+   *   Icon name.
+   * @return string
+   *   Unicode character.
+   */
+  public function getUnicodeChar($name) {
```
Comment (and1truong) [inline lib/Icon/FontEllo.php:49]: This method is too long. Can we split it?

File: lib/Icon/FontEllo.php [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+   *
+   * @staticvar boolean $libraries_added
+   * @staticvar boolean $inline_data_css_added
+   * @staticvar boolean $loading_font_css_added
+   * @param string $name
+   * @param string $unicode_char
+   * @param string $font_path
+   * @param string $font_name
+   */
+  public function getCss($name, $unicode_char, $font_path, $font_name) {
```
Comment (and1truong) [inline lib/Icon/FontEllo.php:147]: This method is too long (complex), can we split?

File: lib/Icon/FontEllo.php [hunk 12]
```diff
@@ -0,0 +1,282 @@
...
+   *   Icon name.
+   * @return string
+   *   Unicode character.
+   */
+  public function getUnicodeChar($name) {
+    $callback = array($this, 'scanDirectory');
+    $directory = at_library('fontello', NULL, FALSE) . 'src';
+    $arguments = array($directory, $name);
+
+    $char = \at_fn::at_cache("atfont:uni_char:{$name}, + 1 year", $callback, $arguments);
```
Comment (and1truong) [inline lib/Icon/FontEllo.php:54]: at_cache('%id:%bin:%ttl') — https://github.com/atdrupal/at_base/wiki/7.x-2.x-Function:-at_cache

**Comment by and1truong** [comment]
@tienvx  please try to make Travis green!

**Comment by and1truong** [comment]
Very close to complete now. Can you check issues reported by https://scrutinizer-ci.com/g/atdrupal/at_base/inspections/464fd7fd-41d9-4a49-a22f-85c576096b10

**Comment by and1truong** [comment]
Code merged in #190

**Pull request closed without merging**